bin/subst.ml

```diff
@@ -299,111 +299,142 @@ let make_watermark_map ~commit ~version ~dune_project ~info =
 
 let subst vcs =
   let open Memo.O in
-  let* (version, commit), files =
-    Memo.fork_and_join
-      (fun () ->
-        Memo.fork_and_join (fun () -> Vcs.describe vcs) (fun () -> Vcs.commit_id vcs))
-      (fun () -> Vcs.files vcs)
-  in
-  let+ (dune_project : Dune_project.t) =
-    (let files =
-       (* Filter-out files form sub-directories *)
-       List.fold_left files ~init:String.Set.empty ~f:(fun acc fn ->
-         let fn = Path.source fn in
-         if Path.is_root (Path.parent_exn fn)
-         then String.Set.add acc (Path.to_string fn)
-         else acc)
-     in
-     Dune_project.load ~dir:Path.Source.root ~files ~infer_from_opam_files:true)
-    >>| function
-    | Some dune_project -> dune_project
-    | None ->
-      User_error.raise
-        ~loc:(Loc.in_dir (Path.source Path.Source.root))
-        [ Pp.text
-            "There is no dune-project file in the current directory, please add one with \
-             a (name <name>) field in it."
-        ]
-        ~hints:
-          [ Pp.concat
-              ~sep:Pp.space
-              [ User_message.command "dune subst"
-              ; Pp.text "must be executed from the root of the project."
-              ]
-            |> Pp.hovbox
-          ]
-  in
-  (let loc, subst_config = Dune_project.subst_config dune_project.project in
-   match subst_config with
-   | `Enabled -> ()
-   | `Disabled ->
-     User_error.raise
-       ~loc
-       [ Pp.concat
-           ~sep:Pp.space
-           [ User_message.command "dune subst"
-           ; Pp.text "has been disabled in this project. Any use of it is forbidden."
-           ]
-       ]
-       ~hints:
-         [ Pp.text
-             "If you wish to re-enable it, change to (subst enabled) in the dune-project \
-              file."
-         ]);
-  let info =
-    let loc, name =
-      match dune_project.name with
+  (match vcs with
+   | Some vcs ->
+     let+ version = Vcs.describe vcs
+     and+ commit_id = Vcs.commit_id vcs
+     and+ files = Vcs.files vcs in
+     Some (version, commit_id, files)
+   | None ->
+     (* We have to do this because scanning the source tree evaluates [-p].
+        That's because [-p] is needed to interpret packages in dune projects
+        correctly. It should not be necessary, so we shuold probably make the
+        package loading lazier. *)
+     Dune_rules.Only_packages.Clflags.set No_restriction;
+     let* root = Source_tree.root () in
+     let project = Source_tree.Dir.project root in
+     if Dune_project.dune_version project < (3, 17)
+     then Memo.return None
```

Comment:
should we print a warning / hint here or would that be too noisy?

Reply:
Too noisy I'd say. Essentially every single package would emit this warning now.